### ISSUE TITLE
Installation issue: py-pytest with external (system) python

### ISSUE THREAD

**Issue by alalazo** [LLNL/spack#27497]
### Steps to reproduce the issue

Create an environment with:
```console
$ mkdir tmp && cd tmp
$ spack env create -d .
$ spack external find python
$ spack add py-pytest
```
and then edit it to have a single external Python. In my case the final `spack.yaml` looks like:
```yaml
# This is a Spack Environment file.
#
# It describes a set of packages to be installed, along with
# configuration settings.
spack:
  # add package specs to the `specs` list
  specs:
  - py-pytest
  view: true
  packages:
    python:
      externals:
      - spec: python@3.8.10+bz2+ctypes+dbm+lzma+nis+pyexpat+readline+sqlite3+ssl~tix~tkinter+uuid+zlib
        prefix: /usr
```
When trying to install the environment you should get errors near the end:
```console
$ spack install
[ ... ]
==> Installing py-py-1.9.0-vohcpazqyqvxez6gopbq45pzxuaundmm
==> No binary for py-py-1.9.0-vohcpazqyqvxez6gopbq45pzxuaundmm found: installing from source
==> Using cached archive: /home/culpo/PycharmProjects/spack/var/spack/cache/_source-cache/archive/9c/9ca6883ce56b4e8da7e79ac18787889fa5206c79dcc67fb065376cd2fe03f342.tar.gz
==> No patches needed for py-py
==> py-py: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 1:
    '/usr/bin/python3.8' '-s' 'setup.py' '--no-user-cfg' 'build'

1 warning found in build log:
  >> 3    WARNING: The wheel package is not available.
     4    Traceback (most recent call last):
     5      File "setup.py", line 45, in <module>
     6        main()
     7      File "setup.py", line 5, in main
     8        setup(
     9      File "/home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-9.3.0/py-setuptools-58.2.0-j3uiv6hk77qtaa3jo2mnek6gke5wzshg/lib/p
          ython3/dist-packages/setuptools/__init__.py", line 152, in setup

See build log for details:
  /tmp/culpo/spack-stage/spack-stage-py-py-1.9.0-vohcpazqyqvxez6gopbq45pzxuaundmm/spack-build-out.txt
[ ... ]
```

### Information on your system

* **Spack:** 0.17.0-171-2f84643c26
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-icelake
* **Concretizer:** clingo

### Additional information

[spack-build-out.txt](https://github.com/spack/spack/files/7551975/spack-build-out.txt)

@adamjstewart 

Note that the issue is just with external Python. Letting Spack build a new interpreter works correctly. Being able to use a system Python is needed to let Spack bootstrap `py-test` and remove the outdated vendored dependency.

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by alalazo** [comment]
As another weird note, the same command seems to execute fine when run manually:
```console
$ spack cd py-pluggy
$ spack build-env py-pluggy -- bash
[ ... ]
```

**Comment by adamjstewart** [comment]
I'm planning on modifying the entire PythonPackage build system to use `pip install` instead of `python setup.py build/install`. I'm hoping that fixes these kinds of issues. Remind me about this issue if you see my PR for this. Will probably get to this during winter break.

**Comment by alalazo** [comment]
We definitely need to do this (I read your https://github.com/pypa/setuptools/discussions/2813 thanks for that!). We probably need to do something with respect to the consistency of the Python packages that might be already installed in a system Python. Right now we just detect the `python` interpreter but we don't detect which packages are installed in it, nor we ensure that we don't try to override part of those packages.

**Comment by adamjstewart** [comment]
Is this fixed by #27798?
